Add helm support

MonkeyCanCode opened the pull request:
# Description

This PR adds helm support when deploying Polaris. This is based off work from https://github.com/projectnessie/nessie/tree/main/helm/nessie. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

This had being tested locally with installation instruction under helm/polaris/README.md

# Checklist:

Please delete options that are not relevant.

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] If adding new functionality, I have discussed my implementation with the community using the linked GitHub issue
- [x] I have signed and submitted the [ICLA](https://github.com/polaris-catalog/polaris/blob/main/ICLA.md) and if needed, the [CCLA](https://github.com/polaris-catalog/polaris/blob/main/CCLA.md). See [Contributing](https://github.com/polaris-catalog/polaris/blob/main/CONTRIBUTING.md) for details.

MonkeyCanCode commented:
@snazy / @flyrain can your guys take a look at this PR for adding helm support?

This is based off work from https://github.com/projectnessie/nessie/tree/main/helm/nessie.

Here are a couple of items for which I would like further clarification and suggestions:
1. Instead of using two k8s services as the current Nessie setup does, I propose using only one k8s service. Nessie uses a headless service for the management port. However, since the pods will be deployed with non-consistent naming (due to k8s deployment), a headless service may not provide significant benefits. There might be a specific purpose for it, so I’d appreciate feedback on this.

2. The current bootstrapped Kind cluster does not support ingress, but this chart does add that support. To allow end-users to use ingress, I will need to update the `kind-registry.sh` script for the Kind cluster definition. Should I proceed with this update?

3. Polaris supports a set of configuration parameters, which I have currently grouped under the `polaris_config` section in `values.yaml`. While this organization might look a bit messy, it makes it easier for end-users to copy and manage their local configurations. What do you think? Are there better ways to organize the configuration for the Polaris service?

4. Some configuration names contain hyphens, so I am using the `index` keyword (e.g., `index .Values "polaris_config" "metaStoreManager" "persistence-unit"`) to handle these cases, based on the assumptions made in item 3.

5. Currently, the image provides only one endpoint for health checks, and the image itself does not include a JSON parser. Therefore, I am using the same endpoint for both `livenessProbe` and `readinessProbe`. Ideally, we should split this endpoint into two distinct endpoints (or create a wrapper endpoint to parse the response) and use them for the respective probes accordingly.

6. If we decide to proceed with Helm support, we may need a separate page or section (e.g., [Providing Secrets for Cassandra](https://projectnessie.org/guides/kubernetes/#providing-secrets-for-cassandra)). Currently, this is not included in this pull request.

flyrain commented:
Thanks @MonkeyCanCode for working on it. Haven't got the change to go through it, but we need the license header in each new files.

MonkeyCanCode commented:
> license header

All good. Added license header.

MonkeyCanCode commented:
> > license header
> 
> All good. Added license header.

It is more common to use `LICENSE` file under helm chart repo. For now, I added license header for each of the files in this PR. Let me know if we should revert and use `LICENSE` file instead.

MonkeyCanCode commented:
Made couple more changes and addressed couple places where I had questions with. Here is the updated list and things I would like to get clarification/suggestions with (along with couple solutions I implemented):

1. Instead of using two k8s services as the current Nessie setup does, I propose using only one k8s service. Nessie uses a headless service for the management port. However, since the pods will be deployed with non-consistent naming (due to k8s deployment), a headless service may not provide significant benefits. There might be a specific purpose for it, so I’d appreciate feedback on this.

2. The current bootstrapped Kind cluster does not support ingress, but this chart does add that support. To allow end-users to use ingress, I will need to update the kind-registry.sh script for the Kind cluster definition. Should I proceed with this update?

3. Polaris supports a set of configuration parameters, which I have currently grouped under the polaris_config section in values.yaml. While this organization might look a bit messy, it makes it easier for end-users to copy and manage their local configurations. What do you think? Are there better ways to organize the configuration for the Polaris service?
```
For this one, I ended up using `persistenceConfigSecret`, `polarisServerConfig`, and `extraEnv`. Here is the reasoning:
- `persistenceConfigSecret` will be use for `persistence.xml` which can contain sensitive info such as password. Thus, I don't have this being part of the helm chart but rather depends on an externally created k8s secret. 
- `polarisServerConfig` will be use for `polaris-server.yml` where we won't be putting sensitive info here. An end-user can quickly update their server config as needed by updating the `values.yaml` for their helm chart. 
- `extraEnv` will be use for handling credential with AWS secret. It has the ability to populate envs from external secret (e.g. an end-user can create external secret then ref to those as env through this...by doing so, the `values.yaml` will be clean and easy for check-in)
```
4. Currently, the image provides only one endpoint for health checks, and the image itself does not include a JSON parser. Therefore, I am using the same endpoint for both livenessProbe and readinessProbe. Ideally, we should split this endpoint into two distinct endpoints (or create a wrapper endpoint to parse the response) and use them for the respective probes accordingly.

5. If we decide to proceed with Helm support, we may need a separate page or section (e.g., [Providing Secrets for Cassandra](https://projectnessie.org/guides/kubernetes/#providing-secrets-for-cassandra)). Currently, this is not included in this pull request.

6. License header for helm chart
```
The standard way for helm chart is to use a single file named `LICENSE` within the root dir of the chart. Let me know if we should still add license header for all files
```

7.  As the persistence.yaml can only be load from a separate jar if not wanting to use the built-in one from resources dir, should we create an init container to pack the config jar on a PV then shared the PV with deployment?
```
This part is taken care as item 3 along with an init container. The init container will take care the config jar creation. The only downside is currently there is no safe guard for bootstrap operation and rerun of bootstrap can wipe the backend. This will be taken care by PR from https://github.com/polaris-catalog/polaris/pull/59. As of now, I put following to ensure it will only run once (unless an end-user remove the helm and reinstall it): 
    "helm.sh/hook": post-install
    "helm.sh/hook-delete-policy": hook-succeeded
```

dimas-b commented on helm/polaris/LICENSE:
```diff
@@ -0,0 +1,13 @@
+  Copyright (c) 2024 Snowflake Computing Inc.
```
Technically this is a copyright statement.

Do we need a separate LICENSE file here in addition to what is in the repo root?

MonkeyCanCode replied:
You are right there too. Yes, I was copied the license header (copyright statement) into this file. I had updated this one as well with the content of https://github.com/polaris-catalog/polaris/blob/main/LICENSE.

Now to answer your question. Assuming we will be publishing this to [Artifact Hub](https://artifacthub.io/) later on. The LICENSE file will be missing in that case with following sample command:
```
helm package polaris/
```

Or we can update our public script to copy the one from root dir then discard it after build. @dimas-b / @flyrain what do you think?

dimas-b replied:
Having a license file in helm charts makes sense to me. I was just trying to avoid duplication... Would a symlink work instead of a copy here?

MonkeyCanCode replied:
Very good idea. Updated.

dimas-b commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,366 @@
+  repository: localhost:5001/polaris
+  # -- The image pull policy.
+  pullPolicy: IfNotPresent
+  # -- Overrides the image tag.
```
nit: why `Overrides`?

MonkeyCanCode replied:
For pullPolicy (in that case "Always")? If we do set it to always, it will need to pull down about 500MB every time.

dimas-b replied:
I simply means that we set (or define) the tag here, we do not "override", right? Is there another property for the tag?

MonkeyCanCode replied:
Oh, I see what you mean. So I set the default tag set on the `Deployment` as well as `Job`. Updated the wording.

dimas-b commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,366 @@
+  # -- Overrides the image tag.
+  tag: "latest"
+  # -- The path to the directory where the polaris-server.yml file should be mounted.
+  configDir: /app/config
```
Is this really user-settable? Could we have a reasonable default for it for all cases? I'd guess it only matters for the actual java command in the pod.

MonkeyCanCode replied:
So the current docker file will mount the config to `/app/polaris-server.yml` as part of image build (so the image itself does contains a default `polaris-server.yml` which is not ideal and use the following command to start the container:
```
CMD ["java", "-jar", "/app/polaris-service-1.0.0-all.jar", "server", "polaris-server.yml"]
```

Then from the deployment template, I am using `configDir` to overwrite the location of `polaris-server.yaml` so as an end-user, I won't need to rebuild the image each time for a single config change with following snippet ([deployment.yaml](https://github.com/polaris-catalog/polaris/blob/4598e61de892783eaab14f6d94294727e5f1f2d3/helm/polaris/templates/deployment.yaml)):
```
command: ["java"]
args: ["-jar", "/app/polaris-service-1.0.0-all.jar", "server", "{{ trimSuffix "/" .Values.image.configDir }}/polaris-server.yml"]
```

And yes, if we can agreed on a default path (which I think we should), I will then remove this one and set it in the `deployment.yaml`. @dimas-b what do you think?

dimas-b replied:
I'd just use `/app/config` everywhere in helm and remove this value (for simplicity).

MonkeyCanCode replied:
Yes, lets use this as config path. Updated.

dimas-b commented on helm/polaris/templates/tests/test-connection.yaml:
```diff
@@ -0,0 +1,36 @@
+{{/*
+  Copyright (c) 2024 Snowflake Computing Inc.
```
nit: why do we have a copyright header in this YAML file, but not in others? Are copyright headers required in Helm for Polaris?

MonkeyCanCode replied:
Good catch on this one. So @flyrain suggested we should have license header for all files we check-in, thus I added them in. After I reviewed the official helm chart best practice and having a separate `LICENSE` file seems to be more suitable for helm chart file. I then removed the license header from all files that I added. This one was a miss. This is removed now.

dimas-b commented on helm/polaris/templates/job.yaml:
```diff
@@ -0,0 +1,87 @@
+      {{- if .Values.persistenceConfigSecret }}
+      initContainers:
+        - name: init-config
+          image: registry.access.redhat.com/ubi9/openjdk-21:1.20-2.1721752936
```
Maybe move this image reference to `values.yaml`?

MonkeyCanCode replied:
> Maybe move this image reference to `values.yaml`?

I like this. Added. Thanks for review.

dimas-b reviewed (commented):
LGTM overall :+1: I have not had a chance to try it in practice yet.. will try to do that soonish... hopefully tomorrow.

dimas-b commented:
@adutra : your feedback would be very welcome too, when you get a chance. thx!

cgpoh commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,374 @@
+          archive: true
+
+  # Either 'jdbc' or 'polaris'; specifies the underlying delegate catalog
+  baseCatalogType: "polaris"
```
question: I have to comment out this line `# baseCatalogType: "polaris"` in order to deploy successfully to my cluster. What is this for?

MonkeyCanCode replied:
I think that is obsolete and got removed from main as part of https://github.com/polaris-catalog/polaris/pull/143/files. I had updated the PR with latest `polaris-server.yml`. Then to answer your question, for any config under `.polarisServerConfig`, you can make change to it as how you will deal with a regular `polaris-server.yml`. As the config changes all the time and I don't want to hard coded the logic in config validation, I am basically reading this section and convert to yaml for configmap. So if some other settings are added/removed later, all you needed to add/remove them accordingly.

cgpoh replied:
Thank you for your explanation and thanks for the helm chart. It's super useful for me to deploy to my cluster :)

MonkeyCanCode replied:
Anytime.

MonkeyCanCode replied:
@dimas-b see above for the issue you reported.

dimas-b commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,370 @@
+
+image:
+  # -- The image repository to pull from.
+  repository: localhost:5001/polaris
```
nit: move this to the top maybe? It's the primary image.

MonkeyCanCode replied:
Done.

dimas-b commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,370 @@
+# Beware that replicas are stateless; don't set this number > 1 when using in-memory meta store manager.
+replicaCount: 1
+
+initImage:
```
nit: I'd call it "toolsImage"... "init" sounds too strong, it feels like this image already contains init logic (which it does not).

MonkeyCanCode replied:
I think init here mainly refers to init container as that is generic in k8s. But yes, we can call it toolsImage.

dimas-b commented on helm/polaris/templates/deployment.yaml:
```diff
@@ -0,0 +1,119 @@
+        - name: init-config
+          image: "{{ tpl .Values.initImage.repository . }}:{{ tpl .Values.initImage.tag . }}"
+          imagePullPolicy: {{ tpl .Values.initImage.pullPolicy . }}
+          command: ['sh', '-c', 'cd /secret && jar -cf /tmp/conf.jar persistence.xml']
```
nit: `jar -C /secret` can probably avoid the `sh` complications

dimas-b replied:
How is `conf.jar` actually used (when it is generated)? Sorry, I could not trace it down :sweat_smile:

MonkeyCanCode replied:
`conf.jar` is the only way as of now to overwrite the `persistenct.xml`. Details in https://github.com/polaris-catalog/polaris/pull/115

MonkeyCanCode replied:
Good point for the `-C` option for `jar` command. Updated and remove dependencies on `cd` and `sh`

dimas-b commented on helm/polaris/templates/deployment.yaml:
```diff
@@ -0,0 +1,119 @@
+          command: ['sh', '-c', 'cd /secret && jar -cf /tmp/conf.jar persistence.xml']
+          volumeMounts:
+            - name: tmp-volume
+              mountPath: /tmp
```
IMHO, it would be nice to avoid mounting as `/tmp`. It's a standard location for temporary files, so mounting an external volume there looks suspicious. Plus, it looks odd that a file we actually care to save and reuse (I assume) is under `/tmp`. By the same logic, the name should probably be something like `conf-volume`. WDYT?

MonkeyCanCode replied:
This path is derived from https://github.com/polaris-catalog/polaris/pull/88. Basically, we need a shared volume between init container and actual runtime container. The init container will then be responsible of generating the config-jar. If we can agreed on a path, I can make the needed change. We can't use `conf-volume` as that is already being used for the `polaris-service.yaml`:
```
          volumeMounts:
            - name: config-volume
              mountPath: /app/config/polaris-server.yml
              subPath: polaris-server.yml
            {{- if .Values.persistenceConfigSecret }}
            - name: tmp-volume
              mountPath: /tmp
            {{- end }}
...
      volumes:
        - name: config-volume
          configMap:
            name: {{ include "polaris.fullname" . }}
```

MonkeyCanCode replied:
Maybe called it db-config-volume then have it mounts on `/app/db-config`?

dimas-b replied:
#88 does not require the config jar to be in `/tmp`. I think any directory name should work (suggestion: `/eclipselink-config`).

Also, in #88 it looks like we still have to set a server config property to point to e.g. `/eclipselink-config/conf.jar`... I do not see any default locations for `/tmp/conf.jar`.

MonkeyCanCode replied:
@dimas-b thanks for checking. Let me try that later tonight. And yes, service config is still needed but not via JAR. This is currently getting populated by `.polarisServerConfig` in the values.yaml.

dimas-b replied:
Yes, but my point is that the location of the generated `conf.jar` should be in the server config too.

dimas-b replied:
Perhaps we can merge the `metaStoreManager.conf-file` property with the other server properties in the ConfigMap.

MonkeyCanCode replied:
so the location of `conf.jar` is in the server config. Here is the how u can set it:
```
type: eclipse-link
persistence-unit: polaris
conf-file: XXXXX
```
Here is the ref: https://github.com/polaris-catalog/polaris/blob/2aab551e53f6c343ea0bab46c443c2da11f0ed2b/extension/persistence/eclipselink/src/main/java/io/polaris/extension/persistence/impl/eclipselink/EclipseLinkPolarisMetaStoreManagerFactory.java#L35

Currently I am not setting validation on the config. I think all we needed here is list this option in the `values.yaml` then make it configurable on the k8s Deployment/Job. What do you think?

dimas-b replied:
From my POV the main idea of having helm charts is to simplify users' life when deploying Polaris to k8s. So, I think we should "tie loose ends" in the sense of making the user configure only what is absolutely required, for example CPU/mem limits, passwords to external systems, etc.

`persistence.xml` is a low-level config file. I do not think users need to deal with it directly. If Eclipselink requires having this file in a jar file, I think it would be preferable to make a (small) java tool (class) for generating `persistence.xml` from higher-level parameters as part of pod startup (not to bother with a separate init container).

I mean: we can make helm work with `persistence.xml` inside a k8s secret as a whole, but is it actually helpful? If I were to have to deploy Polaris I'd prefer to build a custom docker image with all the configs adjusted to my specific needs rather than managing an XML config in a secret.

MonkeyCanCode replied:
I do agreed the current approach for conf.jar is pretty messy due to the limitation of EclipseLink. However, only provide user to enter username/password/jdbc user will probably not be a prod graded (e.g. default num of connection pool for EclipseLink is 32 which is really useful for a catalog). However, I do agreed that is very details and low level. 

I won't think most people will want to make their own docker image (not only because there is no doc for ref to conf.jar and why that is needed) as those will be more complex for them IMO (however, as of now people will need to build their own image due to lack of driver dependencies). The default docker image is really just not prod graded if an end-user tried to use it for anything close to prod upgrade as a simple config change will mean a complete new docker image (e.g. adjust connection pool from 32 to 64).

Then for simplicity, i think we set the default to in-memory which achieved this purpose already?

dimas-b replied:
Starting with in-memory only helm chart SGTM :)

MonkeyCanCode replied:
Sounds good and that is already the case I think with the default value I set in the values.yaml. Unless you want to remove other things such as replication HPA etc to avoid confusion?

MonkeyCanCode replied:
Sounds good and that is already the case I think with the default value I set in the values.yaml. Unless you want to remove other things such as replication HPA etc to avoid confusion?

dimas-b replied:
I'm going to be away for some days... will review later.

MonkeyCanCode replied:
The change from `/tmp` to `/eclipselink-config` is completed and added a note in `values.yaml` regarding not to change it when using custom `persistence.xml` (by default, it will still be in-memory. I think this should be acceptable for now and hopefully later on once devs fix the EclipseLink issues we mentioned above, we can then come back optimize this part. What do you think?

dimas-b commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,370 @@
+#        key: secret_access_key
+
+# -- Configures for persistence.xml (ensure the key name is 'persistence.xml')
+persistenceConfigSecret: polaris-secret
```
We should probably require the user specify a value for this secret since it's not present by default.

MonkeyCanCode replied:
Yes, this one is identically left out and not part of helm chart as it will contains the db auth/connection info for the backend. This is left as default and users can change it to whatever they like. It is not hard-coded on the actual ref on the other end. This is similar to how nessie is handling it:
```
# Cassandra settings. Only required when using CASSANDRA version store type; ignored otherwise.
cassandra:
  keyspace: nessie
  # -- The contact points for the Cassandra cluster. At least one contact point must be provided,
  # but more can be added for redundancy. The format is a comma-separated list of host:port elements.
  contactPoints: cassandra.cassandra.svc.cluster.local:9042
  localDatacenter: datacenter1
  secret:
    # -- The secret name to pull Cassandra credentials from.
    name: cassandra-creds
    # -- The secret key storing the Cassandra username.
    username: cassandra_username
    # -- The secret key storing the Cassandra password.
    password: cassandra_password
```
But if we have a better value to use, I can change it.

dimas-b commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,370 @@
+#        name: aws-secret
+#        key: secret_access_key
+
+# -- Configures for persistence.xml (ensure the key name is 'persistence.xml')
```
nit: putting the entire `persistence.xml` into a secret looks like an overkill to me, but we can refactor that later.

Please add README instructions for creating this secret.

MonkeyCanCode replied:
Yes on the documentation part. Unfortunally, that is the only possible way as of now with EclipseLink unless we want to modify code to pull those auth info from other places such as env variables etc for the properties obj. See details in https://github.com/polaris-catalog/polaris/pull/88

dimas-b replied:
I think ideally the k8s secret should only contain values for `jakarta.persistence.jdbc.user` and `jakarta.persistence.jdbc.password`. Could we generate `persistence.xml` from those secrets (passed in as env. variables) in the init container?

For example: `sed -e "s/@@PASSWORD@@/$PASSWORD_FROM_K8S_SECRET/" <persistence.xml`

We'd have to load a copy of `persistence.xml` in helm and put the `@@PASSWORD@@` placeholder in the appropriate place.

MonkeyCanCode replied:
I am not very familiar with eclipse link and not sure if that is the only way for people to load username/password (e.g. what if the cert or another secret token is needed for connecting to backend) thus left it as an entire config in the jar without doing substitution. Then technically, we can populate the template for `persistence.xml` in `.values.yaml` then give instructions to user on what to set (e.g. `@@PASSWORD@@`) and behinds those onto the config and does the substitution. By doing so, we will not be limited to only allow the substitution of `URL, USERNAME, and PASSWORD`. What do you think?

dimas-b replied:
@MonkeyCanCode : I very much appreciate your work on starting a helm chart for Polaris!.. Still, it feels like the user experience is pretty complicated at this stage. As I commented in another thread, I think it would be nice to make some tooling first to simplify managing Polaris based on k8s-native configuration (simple config values or ConfgMap) and then revisit helm charts. I think that would be more manageable for developers and have better UX. WDYT?

dimas-b replied:
Alternatively, we could limit the scope of helm charts to in-memory operation initially (this PR) and add Eclipselink later.. WDYT?

MonkeyCanCode replied:
@dimas-b anytime. Yes, I do think the EclipseLink is really a mess at this stage and additional tooling for sure will be more helpful. I am open to ideas on how we should approach this. 

Then for limiting the scope, the default value is in-memory and even if users tried to set to EclipseLink, it wont work as there is no driver loaded in the gradle (update of code is needed). However, for advance user, this will be useable IMO as it is really 2-3 lines change and the service will be up and running in their k8s setup with proper persistent backend enabled.

dimas-b replied:
some ideas (beyond the scope of this PRs): [Quarkus](https://quarkus.io/) works pretty well for us in Nessie. It supports lots of backends and has k8s-friendly config.

MonkeyCanCode replied:
Yes, here is a similar discussion: https://github.com/polaris-catalog/polaris/discussions/125

dimas-b reviewed (commented):
I tried running with this helm, but with `persistenceConfigSecret` set to `null` (meant to run "in memory"), the server fails with:
```
INFO  [2024-08-15 21:57:30,430 - 8242  ] [main] [] i.d.j.DiscoverableSubtypeResolver: Unable to load io.polaris.extension.persistence.impl.eclipselink.EclipseLinkPolarisMetaStoreManagerFacto
INFO  [2024-08-15 21:57:31,173 - 8985  ] [main] [] i.d.c.s.DefaultServerFactory: Registering jersey handler with root path prefix: /                                                          
INFO  [2024-08-15 21:57:31,183 - 8995  ] [main] [] i.d.c.s.DefaultServerFactory: Registering admin handler with root path prefix: /                                                           
ERROR [2024-08-15 21:57:31,405 - 9217  ] [main] [] i.p.service.PolarisApplication: Unrecognized baseCatalogType: null                                                                         
java.lang.RuntimeException: Invalid baseCatalogType: null                                                                                                                                     
    at io.polaris.service.PolarisApplication.run(PolarisApplication.java:211)                                                                                                                 
    at io.polaris.service.PolarisApplication.run(PolarisApplication.java:113)                                                                                                                 
    at io.dropwizard.core.cli.EnvironmentCommand.run(EnvironmentCommand.java:66)                                                                                                              
    at io.dropwizard.core.cli.ConfiguredCommand.run(ConfiguredCommand.java:98)                                                                                                                
    at io.dropwizard.core.cli.Cli.run(Cli.java:78)                                                                                                                                            
    at io.dropwizard.core.Application.run(Application.java:94)                                                                                                                                
    at io.polaris.service.PolarisApplication.main(PolarisApplication.java:117)                                                                                                                
```

MonkeyCanCode commented:
> I tried running with this helm, but with `persistenceConfigSecret` set to `null` (meant to run "in memory"), the server fails with:
> 
> ```
> INFO  [2024-08-15 21:57:30,430 - 8242  ] [main] [] i.d.j.DiscoverableSubtypeResolver: Unable to load io.polaris.extension.persistence.impl.eclipselink.EclipseLinkPolarisMetaStoreManagerFacto
> INFO  [2024-08-15 21:57:31,173 - 8985  ] [main] [] i.d.c.s.DefaultServerFactory: Registering jersey handler with root path prefix: /                                                          
> INFO  [2024-08-15 21:57:31,183 - 8995  ] [main] [] i.d.c.s.DefaultServerFactory: Registering admin handler with root path prefix: /                                                           
> ERROR [2024-08-15 21:57:31,405 - 9217  ] [main] [] i.p.service.PolarisApplication: Unrecognized baseCatalogType: null                                                                         
> java.lang.RuntimeException: Invalid baseCatalogType: null                                                                                                                                     
>     at io.polaris.service.PolarisApplication.run(PolarisApplication.java:211)                                                                                                                 
>     at io.polaris.service.PolarisApplication.run(PolarisApplication.java:113)                                                                                                                 
>     at io.dropwizard.core.cli.EnvironmentCommand.run(EnvironmentCommand.java:66)                                                                                                              
>     at io.dropwizard.core.cli.ConfiguredCommand.run(ConfiguredCommand.java:98)                                                                                                                
>     at io.dropwizard.core.cli.Cli.run(Cli.java:78)                                                                                                                                            
>     at io.dropwizard.core.Application.run(Application.java:94)                                                                                                                                
>     at io.polaris.service.PolarisApplication.main(PolarisApplication.java:117)                                                                                                                
> ```

I think that is due to setting of "baseCatalogType" in the values.yaml as the latest code removed that. Same responded to @cgpoh above.

MonkeyCanCode commented:
@flyrain / @dimas-b  AS we are now using symbolic link for `LICENSE` file within helm dir as a way to reduce duplicate file (as the root dir already has one), it will then actually throws following warning:
```
walk.go:74: found symbolic link in path: /home/xxxx/polaris/helm/polaris/LICENSE resolves to /home/xxxx/polaris/LICENSE. Contents of linked file included and used
```

Quick check shows this is intentional: https://github.com/helm/helm/issues/7019

Should we still proceed with the symbolic link approach? Or should we even have this file within helm at all?

dimas-b commented:
> Should we still proceed with the symbolic link approach? Or should we even have this file within helm at all?

As for me, the symlink message does not bother me too much :) It's a build time message. When we publish the helm charts, end users will get a proper LICENSE file from the helm repo, so they should not get this message. I think it's worth including.

If other people find the message annoying, we can always go back to having a copy of the LICENCE file.

MonkeyCanCode commented:
As https://github.com/polaris-catalog/polaris/pull/59 is now merged and I verified it works as expected. The k8s job in this helm is set to run only once so it won't redo bootstrap during upgrade. However, it was possible for an user to delete the helm deployment and redeploy it (which will then trigger the bootstrap job if we don't have changes in https://github.com/polaris-catalog/polaris/pull/59). Now as the functionality for checking if the backend is already bootstrapped, I had added a new flag to allow user to skip the bootstrap job in case they want to skip it (e.g.  user may want to skip the bootstrap job due to accidently delete helm deployment or already have a backend that is already configured etc.). Thanks @eric-maynard for this change.

MonkeyCanCode commented:
As https://github.com/polaris-catalog/polaris/pull/114 changed the jar name, I will update this PR later tonight to reflect this change.

MonkeyCanCode commented:
> As #114 changed the jar name, I will update this PR later tonight to reflect this change.

This is completed.

cgpoh commented:
Thanks @MonkeyCanCode, LGTM!

MonkeyCanCode commented:
> Thanks @MonkeyCanCode, LGTM!

Thanks for review.

dimas-b commented:
@MonkeyCanCode : the latest CI run shows `Apache Rat audit failure - 12 unapproved licenses`

MonkeyCanCode commented:
> @MonkeyCanCode : the latest CI run shows `Apache Rat audit failure - 12 unapproved licenses`

Thanks for the headup. I will check later tonight to see what is going on. Rat was added last time by others. Will check what is being done on the check side and add the needed things on this PR.

MonkeyCanCode commented:
> > @MonkeyCanCode : the latest CI run shows `Apache Rat audit failure - 12 unapproved licenses`
> 
> Thanks for the headup. I will check later tonight to see what is going on. Rat was added last time by others. Will check what is being done on the check side and add the needed things on this PR.

This is completed. So RAT is looking for licensing header for all files. I had added back the license header that we removed in the past.

MonkeyCanCode commented:
@dimas-b @flyrain anything else is needed for this to merge in?

adutra commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,391 @@
+    # polaris-server: The port the Polaris server listens on for API requests.
+    polaris-service: 8181
+    # polaris-mgmt: The port the Polaris server listens on for metrics API requests (health checks, metrics, etc.).
+    polaris-metrics: 8182
```
Should this be `polaris-mgmt` as implied by the comment above?

MonkeyCanCode replied:
Good catch here. Renamed to `polaris-metrics` in both places. This is because for Nessie, the management API does more than just metrics. Then on Polaris side, 8182 only has metrics as of now. See naming discussion in https://github.com/apache/polaris/pull/127.

adutra commented on helm/polaris/templates/service.yaml:
```diff
@@ -0,0 +1,42 @@
+  selector:
+    {{- include "polaris.selectorLabels" . | nindent 4 }}
+  ports:
+    {{- range $portName, $portNumber := .Values.service.ports }}
```
Could be good to separate the metrics port from the main API port into 2 distinct services. But that can be done later.

MonkeyCanCode replied:
For sure beneficial such as network policies and traffic managements. However, I am not seeing much helm deployment separate metrics away from their main API (good practices for sure but most are not doing it). I will be more than happy to do it but here are couple things I would like to get feedback with:
1. Should we put ingress on top of metrics endpoint (I am assuming no here?)
2. Should we support diff service settings for diff endpoints (e.g. sessionAffinity, service type, sessionAffinity, etc.)

adutra replied:
> Should we put ingress on top of metrics endpoint (I am assuming no here?)

No, and that's the main reason why in Nessie we split the services in two. You typically want to expose the main port, but not the admin/mgmt one. So if you need to put an ingress in front of it, or turn the service into a LoadBalancer or NodePort type, the change would affect both ports and potentially expose the mgmt port as well.

> Should we support diff service settings for diff endpoints

Yes, you could have e.g. different session affinities, but more importantly, you can turn the mgmt service into a headless one.

Finally, you can also set `publishNotReadyAddresses: true`. This allows pods to be addressable even if they are not ready, which allows health checks to provide more insights as to why the pod is not healthy.

But I agree to look into this later on.

MonkeyCanCode replied:
Thanks. I can put those in the next PR after current one got merged.

adutra reviewed (commented):
Hi, sorry for the late review as I was out on vacation :-) 
Looks mostly good to me, just a few minor remarks.

adutra commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,391 @@
+        - type: console
+
+        # Settings for logging to a file.
+        - type: file
```
Should file-based logging be enabled by default for K8s deployments? Isn't console enough?

MonkeyCanCode replied:
For this one, we won't need file-based logging for k8s. Should we remove it? Reasoning on putting it there is because that is the complete content of Polaris Server config (https://github.com/apache/polaris/blob/main/polaris-server.yml). I can remove the file appender if that is more preferred?

adutra replied:
I'm fine either way. You could also comment out the file section but leave it in here just in case, wdyt?

MonkeyCanCode commented on helm/polaris/values.yaml:
```diff
@@ -0,0 +1,391 @@
+
+image:
+  # -- The image repository to pull from.
+  repository: localhost:5001/polaris
```
Unfortunally, there is no published docker image as of now: https://github.com/apache/polaris/issues/152

adutra replied:
Oh I wasn't aware – need to catch up 😅 
What we have here is fine then.

MonkeyCanCode replied:
All good. I can help with those docker/helm publishment as well via CD pipeline if we are okay to proceed with it.

MonkeyCanCode commented:
@adutra thanks for taking a look and nothing to worry. I had addressed most of the issues you listed above. I will run some final tests later tonight then push out those changes. For the ones that are not yet addressed, I had put a comment for additional discussion/clarification.

MonkeyCanCode commented:
> @adutra thanks for taking a look and nothing to worry. I had addressed most of the issues you listed above. I will run some final tests later tonight then push out those changes. For the ones that are not yet addressed, I had put a comment for additional discussion/clarification.

@adutra all comments had being addressed and tested locally. Please take another look when you get a chance.

adutra reviewed (approved):
LGTM, thanks @MonkeyCanCode !

dimas-b commented:
Sorry about the delay from my side... I can have another look next week, but please feel free to merge based on @adutra 's review :)

dimas-b reviewed (approved):
I tested this helm chart with "in memory" mode and it worked well.

I did not test the EclipseLink secret. TBH, I do not think it is very user-friendly to put the whole XML file with persistence config into a k8s secret, but I think it is fine to have it like this for a start.

Thanks for your work on this @MonkeyCanCode !

MonkeyCanCode commented:
> I tested this helm chart with "in memory" mode and it worked well.
> 
> I did not test the EclipseLink secret. TBH, I do not think it is very user-friendly to put the whole XML file with persistence config into a k8s secret, but I think it is fine to have it like this for a start.
> 
> Thanks for your work on this @MonkeyCanCode !

Yeah, I agreed that is not the best. We can improve it once we have a better handle of EclipseLink and build tools around to support it.

dimas-b commented:
@snazy @RussellSpitzer : pinging for review/merge, please :)

adutra commented:
@snazy @RussellSpitzer : can you please review and merge 🙏 ?

MonkeyCanCode commented:
Anything else needed on this PR?

flyrain commented:
@MonkeyCanCode, thank you so much for your work on this. Apologies for the delay; we're currently limited by the number of committers. We'll resolve this issue shortly.

MonkeyCanCode commented:
> @MonkeyCanCode, thank you so much for your work on this. Apologies for the delay; we're currently limited by the number of committers. We'll resolve this issue shortly.

All good. Thanks for confirmation. Was planning to rollout more enhancement on this one thus not wanting to add those on top of this PR to avoid of increase of scope.

mrcnc commented on helm/polaris/Chart.yaml:
```diff
@@ -0,0 +1,31 @@
+home: https://polaris.io/
+icon: https://raw.githubusercontent.com/polaris-catalog/polaris/main/docs/img/logos/polaris-brandmark.png
+sources:
+  - https://github.com/polaris-catalog/polaris
```
```suggestion
  - https://github.com/apache/polaris
```

MonkeyCanCode replied:
Fixed. Thanks.

mrcnc commented on helm/polaris/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Source Code
+
+* <https://github.com/polaris-catalog/polaris>
```
```suggestion
* <https://github.com/apache/polaris>
```

MonkeyCanCode replied:
Fixed. Thanks.

RussellSpitzer commented on helm/polaris/README.md:
```diff
@@ -0,0 +1,119 @@
+
+### Optional
+
+When using custom `persistence.xml`, you will need to create a K8S Secret for `.persistenceConfigSecret`. Here is a sample instruction:
```
Don't use "you" in documentation. 

"Custom `persistence.xml` requires a Kubernetes Secret for `.pers...` 

Example Secret Creation:
```bash
...
```

MonkeyCanCode replied:
Fixed

RussellSpitzer commented:
Thanks @MonkeyCanCode and @adutra , @dimas-b , @flyrain, @cgpoh  and @mrcnc  for your reviews!
